docs/architecture/adr-010-posts-module.md

```diff
@@ -0,0 +1,345 @@
+message Url {
+  required uint64 start = 1;
+  required uint64 end = 2;
+  required string url = 3;
+  optional string display_url = 4;
+}
+```
+
+#### Attachments
+
+```protobuf
+// Attachment contains the data of a single post attachment
+message Attachment {
+  required uint32 id = 1;
+  oneof sum {
+    Poll poll = 2;
+    Media media = 3;
+  }
+}
+
+// Media represents a media attachment
+message Media {
+  required string uri = 2;
+  required string mime_type = 3;
+}
+
+// Poll represents a poll attachment
+message Poll {
+  // Question of the poll
+  required string question = 2;
+  
+  // Answers the users can choose from
+  repeated ProvidedAnswer provided_answers = 3;
+  
+  // Date at which the poll will close
+  required google.protobuf.Timestamp end_date = 4;
+  
+  // Whether the poll allows multiple choices from the same user or not 
+  optional bool allows_multiple_answers = 5 [default = false];
+  
+  // Whether the poll allows to edit an answer or not
+  optional bool allows_answer_edits = 6 [default = false];
+  
+  // Provided answer contains the details of a possible poll answer 
+  message ProvidedAnswer {
+    // Text of the answer
+    optional string text = 1;
+    
+    // Attachments of the answer
+    repeated Attachment attachments = 2;
+  }
+}
+```
+
+### `Msg` Service
+We will allow the following operations to be performed.
+
+**Post management**
+- Create a post 
+- Edit an existing post
+- Add an attachment to a post
+- Remove an attachment from a post
+- Delete a post   
+
+**Post interaction**
+- Answer a post poll
+
+> NOTE  
+> In order to make sure subspace moderators and admins can make sure the ToS of their application is always respected, both the removing of an attachment and the deletion of a post should be allowed also to such people.
+
+> NOTE  
+> As per their nature, _attachments_ are **immutable**. This means that the only operations allowed on a post attachments are either adding an attachment or deleting an existing attachment. No edit on the attachment itself is permitted.  
+
+
+
+```protobuf
+service Msg {
+  // CreatePost allows to create a new post
+  rpc CreatePost(MsgCreatePost) returns (MsgCreatePostResponse);
+  
+  // EditPost allows to edit an existing post
+  rpc EditPost(MsgEditPost) returns (MsgEditPostResponse);
+  
+  // AddPostAttachment allows to add a new attachment to a post
+  rpc AddPostAttachment(MsgAddPostAttachment) returns (MsgAddPostAttachmentResponse);
+  
+  // RemovePostAttachment allows to remove an attachment from a post
+  rpc RemovePostAttachment(MsgRemovePostAttachment) returns (MsgRemovePostAttachmentResponse);
+  
+  // DeletePost allows to delete an existing post
+  rpc DeletePost(MsgDeletePost) returns (MsgDeletePostResponse);
+  
+  // AnswerPoll allows to answer a post poll
+  rpc AnswerPoll(MsgAnswerPoll) returns (MsgAnswerPollResponse);
+}
+
+// MsgCreatePost represents the message to be used to create a post.
+message MsgCreatePost {
```

Comment:
Current `MsgCreatePost` does not include `attachments` field. Creating a post with attachments requires two transactions, creating post and adding post attachments. I think we should allow users to create a post with attachments to reduce tx fees.

Reply:
Thanks, added!